homeassistant/components/mqtt/fan.py

```diff
@@ -133,7 +183,23 @@ def __init__(self, hass, config, config_entry, discovery_data):
         self._templates = None
         self._optimistic = None
         self._optimistic_oscillation = None
+        self._optimistic_percentage = None
+        self._optimistic_preset_mode = None
         self._optimistic_speed = None
+        if CONF_SPEED_COMMAND_TOPIC in config and CONF_SPEED_LIST in config:
+            if not isinstance(
+                config[CONF_SPEED_LIST], list
+            ) or not speed_list_without_preset_modes(config[CONF_SPEED_LIST]):
+                raise ValueError(
+                    "No valid speeds configured, speeds mode feature disabled"
+                )
+        if (
+            CONF_PRESET_MODE_COMMAND_TOPIC in config
+            and not config[CONF_PRESET_MODES_LIST]
+        ):
+            raise ValueError(
+                "No valid preset_modes configured, preset mode feature disabled"
+            )
```

Comment:
Please move all of these checks to the schema itself!

1.
This part:
```
if not isinstance(
                config[CONF_SPEED_LIST], list
            )
```
Is the same as `cv.ensure_list`

2.
Call `speed_list_without_preset_modes` from the schema

3.
As mentioned several times now, replace these checks:
```py
CONF_PRESET_MODE_COMMAND_TOPIC in config
            and not config[CONF_PRESET_MODES_LIST]
```
with marking `CONF_PRESET_MODE_COMMAND_TOPIC` and `CONF_PRESET_MODES_LIST` `vol.Inclusive`.

Reply:
The code was updated resolving the validation issues.